frontend/dockerfile/dockerfile2llb/directives.go

```diff
@@ -2,16 +2,30 @@ package dockerfile2llb
 
 import (
 	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
 	"io"
 	"regexp"
 	"strings"
+	"sync"
 
 	"github.com/moby/buildkit/frontend/dockerfile/parser"
 )
 
 const keySyntax = "syntax"
 
-var reDirective = regexp.MustCompile(`^#\s*([a-zA-Z][a-zA-Z0-9]*)\s*=\s*(.+?)\s*$`)
+type directiveParser interface {
+	parse(r io.Reader) map[string]Directive
```

Comment:
This may still be unnecessarily complex. We don't need to parse directive maps in 3 ways. That is only needed for the `reDirective`. For the rest, we only need to find the record for `"syntax"`.

Reply:
I think the code structure as is still makes sense, even if we only want the syntax - in the future we may want to extract other keys.

Though looking back on this after some time, I wonder if we should drop the `ParseDirectives` function entirely, the only place it gets used is in this package - no other buildx/buildkit code calls it, and if an outside user really wants to determine the directives in a Dockerfile, we should export a function from the `frontend/dockerfile/parser` package instead.

Reply:
The code tries to be clever but therefore is harder to follow and doesn't actually do the logical steps (parse directives, if no directives then skip shebang if present and try if syntax is saved in another way.

The directive parsing(not syntax) now allows shebang.

`jsonDirectiveParser` isn't actually a directive parser(as it should not be) because it has hardcoded logic for `keySyntax`.

Iiuc you could use `//` for regular directives in Dockerfile now as long as there isn't `#syntax`.